api/utils/retryutils/retryv2.go

```diff
@@ -0,0 +1,236 @@
+	// could be 0
+	First time.Duration
+	// Driver generates the underlying progression of delays. Cannot be nil.
+	Driver Driver
+	// Max is a maximum value of the progression,
+	// can't be 0
+	Max time.Duration
+	// Jitter is an optional jitter function to be applied
+	// to the delay.  Note that supplying a jitter means that
+	// successive calls to Duration may return different results.
+	Jitter Jitter `json:"-"`
+	// AutoReset, if greater than zero, causes the linear retry to automatically
+	// reset after Max * AutoReset has elapsed since the last call to Incr.
+	AutoReset int64
+	// Clock to override clock in tests
+	Clock clockwork.Clock
+}
+
+// CheckAndSetDefaults checks and sets defaults
+func (c *RetryV2Config) CheckAndSetDefaults() error {
+	if c.Driver == nil {
+		return trace.BadParameter("missing parameter Driver")
+	}
+	if err := c.Driver.Check(); err != nil {
+		return trace.Wrap(err)
+	}
+	if c.Max == 0 {
+		return trace.BadParameter("missing parameter Max")
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	return nil
+}
+
+// NewRetryV2 returns a new instance of linear retry
+func NewRetryV2(cfg RetryV2Config) (*RetryV2, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return newRetryV2(cfg), nil
+}
+
+// newRetryV2 creates an instance of RetryV2 from a
+// previously verified configuration.
+func newRetryV2(cfg RetryV2Config) *RetryV2 {
+	closedChan := make(chan time.Time)
+	close(closedChan)
+	return &RetryV2{RetryV2Config: cfg, closedChan: closedChan}
+}
+
+// RetryV2 is used to calculate retry period
+// that follows the following logic:
+// On the first error there is no delay
+// on the next error, delay is FastRetryV2
+// on all other errors, delay is SlowRetryV2
+type RetryV2 struct {
+	// RetryV2Config is a linear retry config
+	RetryV2Config
+	lastUse    time.Time
+	attempt    int64
+	closedChan chan time.Time
+}
+
+// Reset resets retry period to initial state
+func (r *RetryV2) Reset() {
+	r.attempt = 0
+}
+
+// ResetToDelay resets retry period and increments the number of attempts.
+func (r *RetryV2) ResetToDelay() {
+	r.Reset()
+	r.Inc()
+}
+
+// Clone creates an identical copy of RetryV2 with fresh state.
+func (r *RetryV2) Clone() Retry {
+	return newRetryV2(r.RetryV2Config)
+}
+
+// Inc increments attempt counter
+func (r *RetryV2) Inc() {
+	r.attempt++
+}
+
+// Duration returns retry duration based on state
+func (r *RetryV2) Duration() time.Duration {
+	if r.AutoReset > 0 {
+		now := r.Clock.Now()
+		if now.After(r.lastUse.Add(r.Max * time.Duration(r.AutoReset))) {
+			r.Reset()
+		}
+		r.lastUse = now
+	}
+
+	a := r.First + r.Driver.Duration(r.attempt)
+	if a < 1 {
+		return 0
+	}
+
+	if a > r.Max {
+		a = r.Max
+	}
+
+	if r.Jitter != nil {
+		a = r.Jitter(a)
+	}
+
+	return a
+}
+
+// After returns channel that fires with timeout
+// defined in Duration method, as a special case
+// if Duration is 0 returns a closed channel
```

Comment:
We don't rely on the special case anywhere for `retryutils.Linear`, can we get rid of it?